### PR TITLE
State context

### PR DESCRIPTION
**Pull request by cicdw**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
This PR introduces a concept of "state context" on all states.  Currently only `Running` states have their context updated to include Task tags.

## Why is this PR important?
This simple idea will power many of our features that are about to be introduced.  

**NOTE:** This PR will prevent any deploys to Prefect Cloud until another release is made!

### CODE REVIEW

**Comment by cicdw** [comment]
**cc:** @jlowin

File: src/prefect/serialization/state.py [hunk 1]
```diff
@@ -32,14 +32,19 @@ class BaseStateSchema(ObjectSchema):
     class Meta:
         object_class = state.State
 
+    context = fields.Dict(
+        key=fields.Str(), values=fields.List(fields.Str()), allow_none=True
```
Comment (cicdw) [inline src/prefect/serialization/state.py]: I'm currently making this serializer hyper-specific.  I could relax this if necessary.

Comment (cicdw) [reply src/prefect/serialization/state.py]: Actually now that I've said this out loud I want to change it in this PR brb

Comment (cicdw) [reply src/prefect/serialization/state.py]: Updated to allow for generic JSON values in the dict.

File: src/prefect/serialization/state.py [hunk 2]
```diff
@@ -32,14 +32,17 @@ class BaseStateSchema(ObjectSchema):
...
         object_class = state.State
 
+    context = fields.Dict(key=fields.Str(), values=JSONCompatible(), allow_none=True)
     message = fields.String(allow_none=True)
     _result = Nested(StateResultSchema, allow_none=False, value_selection_fn=get_safe)
 
     @post_load
     def create_object(self, data: dict, **kwargs: Any) -> state.State:
         result_obj = data.pop("_result", result.NoResult)
+        context = data.pop("context", dict())
```
Comment (jlowin) [inline src/prefect/serialization/state.py]: We actually have a good pattern for this (and the above line, which I didn't notice until now).

If you adjust this serializer's `Meta`, it'll automatically exclude attributes from `__init__`. Then you can still assign them as in line 45.

```
class MySerializer:
    class Meta:
        object_class = WhatEver
		exclude_fields = ['context']
```

Comment (cicdw) [reply src/prefect/serialization/state.py]: Hm whenever I do that, I still have to `pop` context out of the data dictionary, or else it an error is raised.  Here's what I updated:

```python
class BaseStateSchema(ObjectSchema):    
    class Meta:                                                                                                                                                            
        object_class = state.State                                                                                                                                        
        exclude_fields = ["context"]                                                                                                                                      
                                                                                                                                                                          
    context = fields.Dict(key=fields.Str(), values=JSONCompatible(), allow_none=True)                                                                                     
    message = fields.String(allow_none=True)                                                                                                                              
    _result = Nested(StateResultSchema, allow_none=False, value_selection_fn=get_safe)                                                                                    
                                                                                                                                                                          
    @post_load                                                                                                                                                            
    def create_object(self, data: dict, **kwargs: Any) -> state.State:                                                                                                    
        result_obj = data.pop("_result", result.NoResult)                                                                                                                 
        data["result"] = result_obj                                                                                                                                       
        base_obj = super().create_object(data)                                                                                                                            
        base_obj.context = data.get("context", dict())                                                                                                                    
        return base_obj   
```

**Comment by cicdw** [comment]
@jlowin wow good catch, I'm sorry I missed such a big oversight!  I convert `tags` to a list when I place them in context now, and updated one my cloud task runner tests to run through the whole serialization step to test against this behavior in the future.

**Pull request merged**